### ISSUE TITLE
Block API

### ISSUE THREAD

**Issue by gziolo** [WordPress/gutenberg#41236]
This overview issue collects an actionable list of impactful tasks that capture possible enhancements to the various [Block APIs](https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/block-api) that exist to power the WordPress block editor.

Block API is intentionally a vague term reflected in the list of action items grouped into categories like block registration, block attributes, block interactions, block modification, etc. What’s in common is that the tasks included aim to resolve the most common friction points or increase the potential of the APIs offered for WordPress block, plugin, and theme authors. In effect, it makes it possible to fulfill the needs of developers wanting to adopt the block editor in their projects.

🛠️ = in development
🤚 = needs to be unblocked

## ❗Top Things

High-impact projects in active development:

- [ ] 🛠️ https://github.com/WordPress/gutenberg/issues/63037
- [x] 🛠️ https://github.com/WordPress/gutenberg/issues/67520
- [x] 🛠️ https://github.com/WordPress/gutenberg/issues/70883
- [ ] 🛠️ https://github.com/WordPress/gutenberg/issues/63001
- [ ] 🛠️ https://github.com/WordPress/gutenberg/issues/64312

## 📖 Backlog

Planned tasks that are being actively researched, waiting for developers to pick them up or on hold until they get unblocked.

### Block Registration

- [ ] 🤚Explore a mechanism improving developer experience to use one API (example: `useBlockProps`, `InnerBlocks`) that run different code depending on the context: edit, save, or frontend (https://github.com/WordPress/gutenberg/pull/48590).
- [ ] Add SlotFill for the block's description in the sidebar to allow custom code (#49887, #49819). At the moment we use block's description which is a simple string. However, it would be sometimes useful to include some custom HTML code like `<a />`.
- [ ] Start using REST API for block types in WordPress core (https://github.com/WordPress/gutenberg/issues/22812).
- [ ] Register multiple blocks from several `block.json` files. At the moment developers have to register every block separately. It could help to update `register_block_type` or add new `register_block_types` on the server to support auto-discovery of multiple blocks through `block.json` in a given target folder.  
- [ ] "Universal blocks" that could be written once and run in both PHP and JS with the help of a compiler (https://github.com/WordPress/gutenberg/discussions/38224).
- [ ] Allow custom path for languages to be defined in `block.json` for translations used with JavaScript ([Trac #54797](https://core.trac.wordpress.org/ticket/54797)).
- [ ] Explore advanced options on the intersection of `ancestor` and `parent` to support allow/deny lists based on the editor, post-type, and template type context (https://github.com/WordPress/gutenberg/issues/46900, https://github.com/WordPress/gutenberg/issues/41062, related discussion starts also in https://github.com/WordPress/gutenberg/issues/37181#issuecomment-1079194941). 
- [ ] Add a warning for block developers if they use reserved words for attribute names (https://github.com/WordPress/gutenberg/issues/48112).
- [ ] Validate each arg before setting `WP_Block_Type` properties (https://github.com/WordPress/gutenberg/issues/48039).
- [x] https://github.com/WordPress/gutenberg/issues/35223

### Block Assets

Block assets are JavaScript (`editorScript`, `script`, `viewScript`) and CSS (`editorStyle`, `style`) defined in `block.json` file.

- [ ] 🤚 https://github.com/WordPress/gutenberg/issues/2768
- [ ] 🛠️ Allow customizations for assets in `block.json` as explained in [Trac #54018](https://core.trac.wordpress.org/ticket/54018#comment:5) and to fix #40447. Tracked in https://github.com/WordPress/gutenberg/issues/46954.
- [ ] Add new utility function to add script / viewScript to an already registered block (#48382). 
- [ ] Iterate on absorbing block styles (currently `__experimentalStyle` in `block.json`) as part of the global styles mechanism (https://github.com/WordPress/gutenberg/issues/45198).
- [ ] https://github.com/WordPress/gutenberg/issues/56838

### Block Attributes

- [ ] 🤚 Solving some semi-related shortcomings with block attributes (https://github.com/WordPress/gutenberg/discussions/49466).
- [ ] 🛠️ Add support for sourced attributes on the server (https://github.com/WordPress/wordpress-develop/pull/6388). [Trac Ticket.](https://core.trac.wordpress.org/ticket/62653)
- [ ] Optional support that optimized blocks for usage as structured data (https://github.com/WordPress/gutenberg/issues/49437).
- [ ] Explore how to connect an individual block attribute with a given UI control and the save flow. In effect, have a simple way to allow editing block attributes without manually writing all code for block controls and inspector controls (https://github.com/WordPress/gutenberg/issues/48245).
- [ ] Exploring how Patterns can be connected to Schemas (#41606).
- [ ] https://github.com/WordPress/gutenberg/issues/65186
- [ ] #57942
- [ ] https://github.com/WordPress/gutenberg/issues/59532

### Block Supports

- [ ] 🤚Allow controlling behaviors/interactions by blocks and themes (https://github.com/WordPress/gutenberg/issues/50029).
- [ ] Ensure that the block editor respects `parent`, `ancestor`, and `supports.multiple` when using drag & drop, or copy & paste (https://github.com/WordPress/gutenberg/issues/53471, https://github.com/WordPress/gutenberg/issues/66645).
- [ ] Add the ability to limit the number of items within an Inner Blocks area (https://github.com/WordPress/gutenberg/issues/42342).
- [ ] https://github.com/WordPress/gutenberg/issues/53989

### Block Variations

- [ ] Add handling for block type name aliases used with block variations (https://github.com/WordPress/gutenberg/issues/43743). They should fall back if necessary to the original block type name, e.g. `woocommerce/product-list` that can get resolved to the `product-list` variation added to the `core/query` block.
- [ ] Add support for the `ancestor` and `parent` settings for the block variation (https://github.com/WordPress/gutenberg/issues/48424).
- [ ] Allow users to select multiple block style variations (https://github.com/WordPress/gutenberg/issues/14598).
- [ ] Allow block styles scoped to variations (https://github.com/WordPress/gutenberg/issues/40621).
- [ ] Add PHP version of `wp.blocks.registerBlockVariation` (https://github.com/WordPress/gutenberg/issues/47170).
- [ ] https://github.com/WordPress/gutenberg/issues/65851

### Block Hooks

- [ ] https://github.com/WordPress/gutenberg/issues/65676

### Content Modelling

- [ ] 🛠️ https://github.com/WordPress/gutenberg/issues/60397
- [ ] 🛠️ https://github.com/WordPress/gutenberg/issues/60954
- [ ] Offer the ability to inject and manipulate inline dynamic tokens for blocks using Rich Text (https://github.com/WordPress/gutenberg/issues/18490 and https://github.com/WordPress/gutenberg/issues/21932).
- [ ] Dynamic replacement of server-provided content in blocks and in HTML attributes (https://github.com/WordPress/gutenberg/discussions/39831).
- [ ] https://github.com/WordPress/gutenberg/issues/64756
- [ ] https://github.com/WordPress/gutenberg/issues/64870

### Various

- [ ] Offer a way to define the list of disallowed blocks to mitigate the limitation of the currently supported allowed list (https://github.com/WordPress/gutenberg/issues/49592, https://github.com/WordPress/gutenberg/issues/38500).
- [ ] https://github.com/WordPress/gutenberg/issues/58233
- [ ] https://github.com/WordPress/gutenberg/issues/62439
- [ ] https://github.com/WordPress/gutenberg/issues/38767
- [ ] 🤚Introduce classification of block validation types to make block deprecation handling more flexible (https://github.com/WordPress/gutenberg/issues/21703).
- [ ] Introducing more robust permission handling across the various capabilities (block registration, locking, etc) so administrators can define what blocks are available for different user roles (or even what capabilities of individual blocks are to be exposed).  See more in [Phase 3: Collaboration, Block Library](https://make.wordpress.org/core/2023/07/10/block-library/).
  - [ ] https://github.com/WordPress/gutenberg/issues/47641.
  - [ ] https://github.com/WordPress/gutenberg/issues/33891.
- [ ] Proper server-side APIs for block modification (https://github.com/WordPress/gutenberg/issues/6494)
- [ ] Expose parsed block data directly through REST API endpoints ([Trac#53603](https://core.trac.wordpress.org/ticket/53603), POC https://github.com/WordPress/gutenberg/pull/2649)
- [ ] Expand `get_posts` $filter parameter to allow a blocks as a return shape ([Trac#53602](https://core.trac.wordpress.org/ticket/53602))
- [ ] Correctly pass parent block context in REST API call for Server-side Render Component (https://github.com/WordPress/gutenberg/issues/40714, https://github.com/WordPress/gutenberg/issues/34882).
- [ ] https://github.com/WordPress/gutenberg/issues/56278
- [ ] Find a way to cover multiblock plugins where you have all of them installed but they wouldn’t get activated until you insert them for the first time. It’s mostly a UX challenge because you could show them in the inserter like blocks from the Block Directory, but they could get higher priority.
- [ ] Better UX to prevent nesting some blocks into self that crashes the editor (the same Reusable Block, Template Part, Post Content, etc).
- [x] "Undo trap": Avoid getting stuck in an editing state (https://github.com/WordPress/gutenberg/issues/8119).
- [ ] Blocks with different versions like Log In/Out block, or different state like Navigation blocks. Explore how to build UI that allows full page control over global states like the user session, or viewport. Feed that information into the block context so users can edit and preview blocks in different contexts (https://github.com/WordPress/gutenberg/issues/57719, see also related discussion https://github.com/WordPress/gutenberg/discussions/38108).
- [ ] Refactor the block parser to work asynchronously (https://github.com/WordPress/gutenberg/pull/15674).

### Developer Experience

- [ ] https://github.com/WordPress/gutenberg/issues/53847
- [ ] https://github.com/WordPress/gutenberg/issues/63906
- [x] https://github.com/WordPress/gutenberg/issues/63026
- [x] https://github.com/WordPress/gutenberg/issues/65614
- [ ] https://github.com/WordPress/gutenberg/issues/15731
- [x] https://github.com/WordPress/gutenberg/issues/43921
- [ ] https://github.com/WordPress/gutenberg/issues/39915

**Comment by luisherranz** [comment]
> Develop strategies for loading JavaScript on the frontend - head vs footer, but also frontend hydration techniques explored in Implement different hydration techniques block-hydration-experiments#12.

This can be a good first step, but acknowledging that we need more than that in the near future: probably `asycn/defer` (something like what @adamsilverstein proposed in https://github.com/WordPress/performance/issues/168) or even do the jump to ESM.

Apart from that, I think we should explore a mechanism to export different code depending on the context (Edit, Save, or Frontend): https://github.com/WordPress/block-hydration-experiments/issues/11

**Comment by gziolo** [comment]
Thank you @luisherranz for your feedback. I included your suggestions together with references.

I also added two more items that I missed initially but have been following closely for a long time:
- Introduce classification of block validation types to make block deprecation handling more flexible (https://github.com/WordPress/gutenberg/issues/21703).
- "Universal blocks" that could be written once and run in both PHP and JS with the help of a compiler (https://github.com/WordPress/gutenberg/discussions/38224).

**Comment by fabiankaegy** [comment]
@gziolo I have reviewed the items that are already tracked here and I myself have created 3 issues recently that I think could all be added: 
- https://github.com/WordPress/gutenberg/issues/42548 (the one you added the comment to)
- https://github.com/WordPress/gutenberg/issues/42342
- https://github.com/WordPress/gutenberg/issues/38958

**Comment by gziolo** [comment]
## ☑️ Completed in WordPress 6.1 cycle

### Block Registration

- Reference PHP file necessary for the block from `block.json`. It could be a new setting `render` in `block.json` ([Trac#53148](https://core.trac.wordpress.org/ticket/53148), https://github.com/WordPress/wordpress-develop/pull/3222, https://github.com/WordPress/gutenberg/pull/42430).
- Better imports for core blocks in the Gutenberg repository and `@wordpress/block-library` to make it easier to consume individual blocks (https://github.com/WordPress/gutenberg/pull/42258).
- Combine block metadata during the build process into one file to avoid loading too many `block.json` files for core blocks (see https://github.com/WordPress/wordpress-develop/pull/2252).

### Block Assets

Some of the details were tracked in https://github.com/WordPress/gutenberg/issues/33542.

- Add multiple assets per type (`editorScript`, `script`) to bring complete consistency to what we already have for styles (https://github.com/WordPress/wordpress-develop/pull/3108). 
- Change the default behavior for `viewScript` when the `render_callback` is present in the block definition ([Trac#56470](https://core.trac.wordpress.org/ticket/56470)).
- Allow plugins to modify a theme.json file in a similar way themes do (https://github.com/WordPress/gutenberg/issues/41707).

### Documentation 

- Cover optional block variations usage in `block.json` (https://github.com/WordPress/gutenberg/issues/40490, https://github.com/WordPress/gutenberg/pull/42539).

**Comment by gziolo** [comment]
## ☑️ Completed in WordPress 6.2 cycle

### Block Modification

- WordPress 6.2: Injecting dynamic data to block HTML markup in PHP (https://github.com/WordPress/gutenberg/pull/42485, https://make.wordpress.org/core/2022/08/19/a-new-system-for-simply-and-reliably-updating-html-attributes/).

**Comment by luisherranz** [comment]
Mentioning this in case you think it'd be good to include it:

- https://github.com/WordPress/gutenberg/discussions/38108

It's related to this item, but I wasn't sure it was talking about the same use case:

> Explore how to build UI that allows full page control over global state like the user session and feed it into the block context so users can preview blocks in different contexts.

**Comment by gziolo** [comment]
> Mentioning this in case you think it'd be good to include it:
> 
> * [How should we handle different layers/states for the same block in the editor? #38108](https://github.com/WordPress/gutenberg/discussions/38108)
> 
> It's related to this item, but I wasn't sure it was talking about the same use case:

Yes, it was the same item so I linked it and rephrased as:

> Blocks with different versions like Log In/Out block, or different states like Navigation block. Explore how to build UI that allows full page control over global states like the user session, or viewport. Feed that information into the block context so users can edit and preview blocks in different contexts.

Unrelated to that I added three action items based on recently triaged issues:
- Add a new utility function to add script / viewScript to an already registered block (https://github.com/WordPress/gutenberg/issues/48382).
- Add support for the ancestor and parent settings for the block variation (#48424).
- Clarify the behavior of the `parent` when empty during block registration (https://github.com/WordPress/gutenberg/issues/15731).

**Comment by gziolo** [comment]
Another set of updates applied:

- Added to Block Variations: Allow users to select multiple block style variations (https://github.com/WordPress/gutenberg/issues/14598).
- Added to Block Variations: Block Variations: Add PHP version of `wp.blocks.registerBlockVariation` (https://github.com/WordPress/gutenberg/issues/47170).
- Replaced in Documentation ~Refresh [README](https://www.npmjs.com/package/@wordpress/blocks) for @wordpress/blocks, which is outdated.~ with: Address confusing experiences when searching resources related to learning about block development (https://github.com/WordPress/gutenberg/issues/48638).
- Added to Documentation: `spacing.blockGap` explanation is inadequate for custom blocks (https://github.com/WordPress/gutenberg/issues/43921).
- Removed as it seems to be not general enough and useful only for a few core blocks: Allow setting role HTML attributes on blocks (#46002).

**Comment by oandregal** [comment]
Sharing in case it relates to this effort and want it tracked here: one of the big tasks related to styles is being able to absorb block styles (CSS as well as the `__experimentalStyle` from `block.json`) as part of the global styles merge algorithm https://github.com/WordPress/gutenberg/issues/45198

This is: by providing block styles in a structured way, they can be merged with core, theme, and user styles into a single stylesheet. Then, decide what of those should be shipped to the front-end.

**Comment by gziolo** [comment]
> Sharing in case it relates to this effort and want it tracked here: one of the big tasks related to styles is being able to absorb block styles (CSS as well as the `__experimentalStyle` from `block.json`) as part of the global styles merge algorithm #45198

That's a great one. I included it for now in the "Block Assets" section as it sort of removes the need to write specific regular CSS. This is also going to be another helpful performance optimization.

**Comment by gziolo** [comment]
Added new item in the Block Modification section:

- Offer the ability to inject and manipulate inline dynamic tokens for blocks using Rich Text (https://github.com/WordPress/gutenberg/issues/18490 and https://github.com/WordPress/gutenberg/issues/21932).

**Comment by gziolo** [comment]
Added  new item in the Block Attributes section:

- Optional support that optimizes handling blocks for usage as structured data (https://github.com/WordPress/gutenberg/issues/49437).

**Comment by jeherve** [comment]
> Consider adding SlotFill for the block's description in the sidebar to allow custom code. At the moment we use block's description which is a simple string. However, it would be sometimes useful to include some custom HTML code like `<a />`.

Now tracked in #49887

**Comment by gziolo** [comment]
> > Consider adding SlotFill for the block's description in the sidebar to allow custom code. At the moment we use block's description which is a simple string. However, it would be sometimes useful to include some custom HTML code like `<a />`.
> 
> Now tracked in #49887

Thank you so much @jeherve. I linked the issue in the description.

---

I also removed the following item from the Block Attributes section:

- Check whether we still need a special way or if it's only a documentation effort to expose attributes for the client in the frontend context with Interactive API in place (related to https://github.com/WordPress/block-hydration-experiments/issues/13).

There is an alternative approach that allows exposing attributes to the client covered in the [Proposal: The Interactivity API – A better developer experience in building interactive blocks](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/).

---

I added a new item in the Block Attributes section:

- Provide a general solution to customize the list of allowed blocks for inner blocks with block attributes (https://github.com/WordPress/gutenberg/issues/15682).

**Comment by gziolo** [comment]
I included a list of the current Top Things:

- 🛠️ https://github.com/WordPress/gutenberg/issues/49868
- 🛠️ https://github.com/WordPress/gutenberg/issues/50029
- 🛠️ https://github.com/WordPress/gutenberg/issues/44410
- 🛠️ https://github.com/WordPress/gutenberg/issues/48458
- 🛠️ https://github.com/WordPress/gutenberg/issues/39439

I also merged the "Backlog" section with "In Progress" section.

**Comment by gziolo** [comment]
I added a new item to the Backlog in the Block Interactions section, it's in progress:

-  Offer a way to ungroup blocks with a new type of transform (https://github.com/WordPress/gutenberg/pull/50385)

**Comment by gziolo** [comment]
In https://github.com/WordPress/gutenberg/discussions/51005, @Mamaduka is championing the proposal to create a formal API for registering new [Block Supports](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/) options for blocks.

**Comment by wojtekLs** [comment]
FUCK

**Comment by wojtekLs** [comment]
Wow, at first I thought that Gutenberg introduces user tracking.............

**Comment by draganescu** [comment]
I found myself wanting for blocks to be able to export data they have. Case in point: trying to implement a feature which shows in the site editor sidebar the navigations available in a template part, when that template part is visible in the frame. I can get the client ids of the navigation blocks, but then I have to reimplement the whole logic of getting navigation details from the API although the blocks rendered already did it. I want to ask client id X give me data :)

**Comment by mtias** [comment]
Added https://github.com/WordPress/gutenberg/issues/51373 to the list.

**Comment by noisysocks** [comment]
Something that's been on my mind lately is how we have a lot of different APIs to do with restricting blocks:

- The `templateLock` prop on `InnerBlocks` and `templateLock` block editor setting
- The `lock` block attribute 
- The `allowedBlocks` prop on `InnerBlocks` and `allowedBlocks` block editor setting
- The `parent` attribute in `registerBlockType`
- The `useBlockEditingMode` hook

I wonder if we can consolidate some of these APIs or otherwise make them more coherent. Perhaps something to think about in the context of Block API improvements.

**Comment by gziolo** [comment]
# Update

## Top Things

- https://github.com/WordPress/gutenberg/issues/53740
- https://github.com/WordPress/gutenberg/issues/53987
- [**HTML API: Introduce HTML Processor**](https://core.trac.wordpress.org/ticket/58517)
- https://github.com/WordPress/gutenberg/issues/54536
- https://github.com/WordPress/gutenberg/issues/48458
- https://github.com/WordPress/gutenberg/issues/51373
- https://github.com/WordPress/gutenberg/issues/2768

## Completed during WP 6.3 & 6.4 

- https://github.com/WordPress/gutenberg/pull/48286
- https://github.com/WordPress/gutenberg/issues/39439
- https://github.com/WordPress/gutenberg/issues/49868
- https://github.com/WordPress/gutenberg/issues/50866
- https://github.com/WordPress/gutenberg/issues/51056
- https://github.com/WordPress/gutenberg/issues/44410
- Stabilize experimental selectors API as top-level `selectors` field ([Trac #57585](https://core.trac.wordpress.org/ticket/57585), https://github.com/WordPress/gutenberg/pull/46496).
- Develop strategies for loading JavaScript on the frontend - head vs footer, async/defer (related proposal https://github.com/WordPress/performance/issues/168). It could also cover frontend hydration techniques explored in https://github.com/WordPress/block-hydration-experiments/issues/12.
- No simple way to enqueue styles and scripts for a group of blocks (https://github.com/WordPress/gutenberg/issues/41821).
- Offer a way to ungroup blocks with a new type of transform (https://github.com/WordPress/gutenberg/pull/50385).
- Docs: Address confusing experiences when searching resources related to learning about block development (https://github.com/WordPress/gutenberg/issues/48638).
- Docs: How to work with the templates defined with `render` (https://github.com/WordPress/gutenberg/issues/47826).

## On hold

- https://github.com/WordPress/gutenberg/issues/50029

## Added

- Introducing more robust permission handling across the various capabilities (block registration, locking, etc) so administrators can define what blocks are available for different user roles (or even what capabilities of individual blocks are to be exposed). See [Phase 3: Collaboration, Block Library](https://make.wordpress.org/core/2023/07/10/block-library/).
- Offer a way to define the list of disallowed blocks to mitigate the limitation of the currently supported allowed list (https://github.com/WordPress/gutenberg/issues/49592, https://github.com/WordPress/gutenberg/issues/38500).
- https://github.com/WordPress/gutenberg/issues/50229
- https://github.com/WordPress/gutenberg/issues/54112
- https://github.com/WordPress/gutenberg/issues/53847

**Comment by gaambo** [comment]
Regarding **`viewStyle`**: The initial issue was closed, therefore I'm posting this here:
I actually think this would be great to have. Not having this assumes, frontend styles can be used 1:1 in the editor. I've encountered many situations where that's just not true - it has gotten a lot better with the `apiVersion: 2` blocks, but there are still many cases where I want to enqueue a style only in the frontend. Examples:
- Very complex blocks - where I will develop a very specific block editing experience in the editor with components and editor styles and want to load the "real" styles for the (dynamically) rendered block only in the frontend. 
- JavaScript-heavy frontend - the most basic example being slider. I will have a lot of CSS which only styles the JS-initialized version of the block and they just don't need to be loaded in the editor. 
- There are cases where the frontend styles are just not used in the editor (and I know that while developing), so it's a performance slow-down if you have many such blocks. But there are also cases where the frontend styles may just "destroy" the editing experience of a block because they just don't work in that context.

Since we already have `script`, `viewScript` and `editorScript` I think it would keep clarity by also introducing `viewStyle`. Even if there are not many use cases where it's used, I don't see a real disadvantage. 
It would also mean, that all assets of a block in all contexts could be managed through `block.json` - which would be great.

**Comment by gziolo** [comment]
@gaambo, thank you for feedback.

> Regarding viewStyle: The initial issue was closed, therefore I'm posting this here:

Can you share the link to the issue? I would like to learn why it was closed and potentially reopen. There is still an item  listed in the Backlog that signals that we have this API on the roadmap but it doesn't link to the issue you referred to:

"Introduce `viewStyle` for styles used only on the front end to align with scripts and `viewScript`."

**Comment by gaambo** [comment]
> Can you share the link to the issue? I would like to learn why it was closed and potentially reopen. There is still an item listed in the Backlog that signals that we have this API on the roadmap but it doesn't link to the issue you referred to:

@gziolo 
I've found #33542 where `viewStyle` was [mentioned](https://github.com/WordPress/gutenberg/issues/33542#issuecomment-1246649599). I can open a new issue if required and we can link it in this overview issue.

**Comment by gziolo** [comment]
> I've found #33542 where `viewStyle` was [mentioned](https://github.com/WordPress/gutenberg/issues/33542#issuecomment-1246649599). I can open a new issue if required and we can link it in this overview issue.

It was me who closed the issue with the note I moved the item to this issue - the one I quoted in https://github.com/WordPress/gutenberg/issues/41236#issuecomment-1720161563 😄 

If you could open a new issue that would be great, thank you! I'm happy to see this feature implemented if it's going to be useful for custom blocks.

**Comment by gziolo** [comment]
# Update

## Completed during WP 6.4

- https://github.com/WordPress/gutenberg/issues/53987


## Added

- https://github.com/WordPress/gutenberg/issues/54904 (**Top Things**)
- https://github.com/WordPress/gutenberg/issues/35223

## Needs review

- Introduce `viewStyle` for styles used only on the front end to align with scripts and `viewScript` (https://github.com/WordPress/gutenberg/pull/55492, https://github.com/WordPress/wordpress-develop/pull/5531).

**Comment by gziolo** [comment]
## Update

### Completed during WordPress 6.5 cycle

- https://github.com/WordPress/gutenberg/issues/53300
- https://github.com/WordPress/gutenberg/issues/56803
- https://github.com/WordPress/gutenberg/issues/54904
- https://github.com/WordPress/gutenberg/issues/53705
- https://github.com/WordPress/gutenberg/discussions/54579
- Create a system for inserting site/post metadata into generic blocks (https://github.com/WordPress/gutenberg/issues/37753.
- Provide a general solution to customize the list of allowed blocks for inner blocks with block attributes (https://github.com/WordPress/gutenberg/issues/15682, https://github.com/WordPress/gutenberg/issues/16560, https://github.com/WordPress/gutenberg/issues/14515).
- Start using JavaScript Modules and Import Maps (https://github.com/WordPress/gutenberg/issues/36716).
- Support for deferred block variation initialization on the server has been added, which improves performance (https://core.trac.wordpress.org/ticket/59969).
- Introduce view style for styles used only on the front end to align with scripts and view script (https://github.com/WordPress/gutenberg/issues/54491, https://core.trac.wordpress.org/ticket/59673).
- Asset registration file .asset.php has been made optional for blocks (https://core.trac.wordpress.org/ticket/60460).
- https://github.com/WordPress/gutenberg/issues/56305

### New

- 🔝 https://github.com/WordPress/gutenberg/issues/60956
- 🔝 https://github.com/WordPress/gutenberg/issues/59819
- 🔝 https://github.com/WordPress/gutenberg/issues/60219
- 🔝 https://github.com/WordPress/gutenberg/issues/60324
- 🔝 https://github.com/WordPress/gutenberg/issues/60252
- https://github.com/WordPress/gutenberg/issues/56838
- https://github.com/WordPress/gutenberg/issues/58233

### Resources

- [Dev note: Interactivity API](https://make.wordpress.org/core/2024/03/04/interactivity-api-dev-note/)
  - [An introduction to block-based mega menus](https://developer.wordpress.org/news/2024/02/29/an-introduction-to-block-based-mega-menus/)
- [Dev note: The Block Bindings API](https://make.wordpress.org/core/2024/03/06/new-feature-the-block-bindings-api/).
  - [Introducing Block Bindings, part 1: connecting custom fields](https://developer.wordpress.org/news/2024/02/20/introducing-block-bindings-part-1-connecting-custom-fields/).
  - [Introducing Block Bindings, part 2: Working with custom binding sources](https://developer.wordpress.org/news/2024/03/06/introducing-block-bindings-part-2-working-with-custom-binding-sources/).
- [Dev note: Updates to the HTML API](https://make.wordpress.org/core/2024/03/04/updates-to-the-html-api-in-6-5/) 
- [Dev note: Updates to Block Hooks ](https://make.wordpress.org/core/2024/03/04/updates-to-block-hooks-in-6-5/)
- [Dev note: Script Modules in 6.5](https://make.wordpress.org/core/2024/03/04/script-modules-in-6-5/)
- [Dev note: Performance improvements for registering block type variations with callbacks](https://make.wordpress.org/core/2024/02/29/performance-improvements-for-registering-block-variations-with-callbacks/)
  - [How to register block variations with PHP](https://developer.wordpress.org/news/2024/03/14/how-to-register-block-variations-with-php/)
- [Tutorial: Build your first block](https://developer.wordpress.org/block-editor/getting-started/tutorial/)

**Comment by gziolo** [comment]
## Update

## Completed during WordPress 6.6 cycle

- https://github.com/WordPress/gutenberg/issues/60956
- https://github.com/WordPress/gutenberg/issues/59819
- https://github.com/WordPress/gutenberg/issues/60219
- https://github.com/WordPress/gutenberg/issues/60324
- https://github.com/WordPress/gutenberg/issues/60252
- https://github.com/WordPress/gutenberg/pull/62088

## Completed during WordPress 6.7 cycle

- https://github.com/WordPress/gutenberg/issues/63018
- https://github.com/WordPress/gutenberg/issues/60396
- https://github.com/WordPress/gutenberg/issues/63232
- https://github.com/WordPress/gutenberg/issues/50027
- https://github.com/WordPress/gutenberg/issues/50229
- https://github.com/WordPress/gutenberg/issues/54112
- https://github.com/WordPress/gutenberg/pull/62092
- https://github.com/WordPress/gutenberg/pull/65484
- https://github.com/WordPress/wordpress-develop/pull/7303
- https://github.com/WordPress/wordpress-develop/pull/7443

## New

- https://github.com/WordPress/gutenberg/issues/63037
- https://github.com/WordPress/gutenberg/issues/63001
- https://github.com/WordPress/gutenberg/issues/64312
- https://github.com/WordPress/gutenberg/issues/67276
- https://github.com/WordPress/gutenberg/issues/67520

## Resources

- [Editing custom fields from connected blocks](https://make.wordpress.org/core/2024/06/28/editing-custom-fields-from-connected-blocks/)
- [Updates to the HTML API in 6.6](https://make.wordpress.org/core/2024/06/24/updates-to-the-html-api-in-6-6/)
- [Updates to the Interactivity API in 6.6](https://make.wordpress.org/core/2024/06/28/updates-to-the-interactivity-api-in-6-6/)
- [Improvements to active block variation detection](https://make.wordpress.org/core/2024/06/18/improvements-to-active-block-variation-detection/)
- [Block Bindings: Improvements to the Editor Experience in 6.7](https://make.wordpress.org/core/2024/10/21/block-bindings-improvements-to-the-editor-experience-in-6-7/)
- [Updates to Block Hooks in WordPress 6.7](https://make.wordpress.org/core/2024/10/21/updates-to-block-hooks-in-wordpress-6-7/)
- [New block type registration APIs to improve performance in WordPress 6.7](https://make.wordpress.org/core/2024/10/17/new-block-type-registration-apis-to-improve-performance-in-wordpress-6-7/)
- [Updates to the HTML API in 6.7](https://make.wordpress.org/core/2024/10/17/updates-to-the-html-api-in-6-7/)

**Comment by gziolo** [comment]
- Issue https://github.com/WordPress/gutenberg/issues/67276 was closed in favor of https://github.com/WordPress/gutenberg/issues/70883.

**Comment by mtias** [comment]
@gziolo I'd like to include this one here https://github.com/WordPress/gutenberg/issues/65676

**Comment by cbravobernal** [comment]
> [@gziolo](https://github.com/gziolo)[<img alt="" width="12" height="12" src="chrome-extension://hgomfjikakokcbkjlfgodhklifiplmpg/images/wp-logo.png">](https://profiles.wordpress.org/gziolo) I'd like to include this one here [#65676](https://github.com/WordPress/gutenberg/issues/65676)

@mtias , added to a new Block Hooks section.
